lua/telescope/pickers/layout_strategies.lua

```diff
@@ -17,129 +17,249 @@
----     - See |layout_strategies.flex|
----
---- Available tweaks to the settings in layout defaults include
---- (can be applied to horizontal and vertical layouts):
----   - mirror (default is `false`):
----     - Flip the view of the current layout:
----       - If using horizontal: if `true`, swaps the location of the
----         results/prompt window and preview window
----       - If using vertical: if `true`, swaps the location of the results and
----         prompt windows
----
----   - width_padding:
----     - How many cells to pad the width of Telescope's layout window
----
----   - height_padding:
----     - How many cells to pad the height of Telescope's layout window
----
----   - preview_width:
----     - Change the width of Telescope's preview window
----
----   - scroll_speed:
----     - Change the scrolling speed of the previewer
 ---@brief ]]
 
-local config = require('telescope.config')
-local resolve = require("telescope.config.resolve")
-
+local resolve = require('telescope.config.resolve')
 local p_window = require('telescope.pickers.window')
 
+local get_border_size = function(opts)
+  if opts.window.border == false then
+    return 0
+  end
+
+  return 1
+end
+
+local layout_strategies = {}
+layout_strategies._configurations = {}
+
+--@param strategy_config table: table with keys for each option for a strategy
+--@return table: table with keys for each option (for this strategy) and with keys for each layout_strategy
+local get_valid_configuration_keys = function(strategy_config)
+  local valid_configuration_keys = {
+    -- TEMP: There are a few keys we should say are valid to start with.
+    preview_cutoff = true,
+    prompt_position = true,
+  }
+
+  for key in pairs(strategy_config) do
+    valid_configuration_keys[key] = true
+  end
+
+  for name in pairs(layout_strategies) do
+    valid_configuration_keys[name] = true
+  end
 
--- Check if there are any borders. Right now it's a little raw as
--- there are a few things that contribute to the border
-local is_borderless = function(opts)
-  return opts.window.border == false
+  return valid_configuration_keys
 end
 
+--@param strategy_name string: the name of the layout_strategy we are validating for
+--@param configuration table: table with keys for each option available
+--@param values table: table containing all of the non-default options we want to set
+--@param default_layout_config: table with the default values to configure layouts
+--@return table: table containing the combined options (defaults and non-defaults)
+local function validate_layout_config(strategy_name, configuration, values, default_layout_config)
+  assert(strategy_name, "It is required to have a strategy name for validation.")
+  local valid_configuration_keys = get_valid_configuration_keys(configuration)
+
+  -- If no default_layout_config provided, check Telescope's config values
+  default_layout_config = default_layout_config or require('telescope.config').values.layout_config
```

Comment:
For stuff like that we should probably slowly move to `vim.F.if_nil` Just to make it even more robust